backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/objcexport/ObjCExportHeaderGenerator.kt

```diff
@@ -909,7 +987,19 @@ abstract class ObjCExportHeaderGenerator internal constructor(
     }
 
     internal fun generateClass(descriptor: ClassDescriptor) {
+        //If type args reference this class in parent declaration, we have a special case forward class declaration
+        //needed. If upper bounds are added back to type parameters, this may need to be expanded.
+        fun selfReferencingClassType(descriptor: ClassDescriptor): Boolean {
+            val parentType = computeSuperClassType(descriptor)
+            return parentType != null && parentType.arguments.any { descriptor == it.type.getErasedTypeClass() }
```

Comment:
It is not the only case.
Please try disabling this optimization instead:
https://github.com/JetBrains/kotlin-native/blob/3a0be1c3319a44fcf8a9d5764fb581a3e268ba5a/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/objcexport/ObjCExportHeaderGenerator.kt#L1018
(i.e. always add forward declaration).
I find it a better option. I would enable it back after some refactoring if required.